pkg/objstore/caching/readers.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package caching
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	tsdberrors "github.com/prometheus/prometheus/tsdb/errors"
+
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// Gives access to sequentially read blocks. Offsets are aligned to blockSize.
+type blocksProvider interface {
+	blockAt(offset int64) ([]byte, error)
+	close() error
+}
+
+// io.Reader implementation that uses blocks from blocksProvider.
+type blocksReader struct {
+	blockSize int64
+	blocks    blocksProvider
+
+	// Offset for next read, used to find correct block to return data from.
+	readOffset int64
+
+	// Remaining data to return from this reader. Once zero, this reader reports EOF.
+	remaining int64
+}
+
+func newBlocksReader(blockSize int64, blocks blocksProvider, readOffset, remaining int64) *blocksReader {
+	return &blocksReader{
+		blockSize:  blockSize,
```

Comment:
In other places it's called `chunkBlockSize`.

Reply:
I see `blocksReader` as being generic reader implementation, while `chunkBlockSize` is related to chunks.